### PR TITLE
feat(utxo-coredao): initialize utxo-coredao

### PR DESCRIPTION
**Pull request by davidkaplanbitgo**
This is dependent on wasm-miniscript with OP_DROP, which is not merged in yet. I will add tests for this once we can import it.

This contains code for the `OP_RETURN` as well as the descriptor for CoreDao

TICKET: BTC-1578

### CODE REVIEW

File: modules/utxo-coredao/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+```shell
+npm i @bitgo/sdk-api @bitgo/utxo-coredao>
+```
+
+## Development
+
+Most of the coin implementations are derived from `@bitgo/sdk-core`, `@bitgo/statics`, and coin specific packages. These implementations are used to interact with the BitGo API and BitGo platform services.
+
+You will notice that the basic version of common class extensions have been provided to you and must be resolved before the package build will succeed. Upon initiation of a given SDK coin, you will need to verify that your coin has been included in the root `tsconfig.packages.json` and that the linting, formatting, and testing succeeds when run both within the coin and from the root of BitGoJS.
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/README.md:19]:
```suggestion
```

File: modules/utxo-coredao/src/transaction.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  validator,
+  fee,
+  redeemScript,
+  timelock,
+}: {
+  version: number;
+  chainId: string;
+  delegator: string;
+  validator: string;
+  fee: string;
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:42]:
```suggestion
  fee: number;
```

File: modules/utxo-coredao/src/transaction.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  delegator,
+  validator,
+  fee,
+  redeemScript,
+  timelock,
+}: {
+  version: number;
+  chainId: string;
+  delegator: string;
+  validator: string;
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:41]:
```suggestion
  chainId: Buffer;
  delegator: Buffer;
  validator: Buffer;
```

File: modules/utxo-coredao/src/transaction.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  fee: string;
+  redeemScript?: string;
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:43]:
```suggestion
  redeemScript?: Buffer;
```

File: modules/utxo-coredao/src/transaction.ts (hunk 3 above)
Comment (OttoAllmendinger) [reply modules/utxo-coredao/src/transaction.ts:41]: let the caller do the encoding

File: modules/utxo-coredao/src/transaction.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  timelock?: number;
+}): Buffer {
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:45]: :+1: let's use the same types in the arguments

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-coredao/src/transaction.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  if (!redeemScript && !timelock) {
+    throw new Error('Either redeemScript or timelock must be provided');
+  }
+  const redeemScriptBuffer = redeemScript ?? Buffer.from([]);
+  if (timelock && (timelock < 0 || timelock > 4294967295)) {
+    throw new Error('Invalid timelock - out of range');
+  }
+  const timelockBuffer = timelock ? Buffer.alloc(4, timelock) : Buffer.from([]);
```
Comment (davidkaplanbitgo) [inline modules/utxo-coredao/src/transaction.ts:82]: Do I need to reverse this buffer?

File: modules/utxo-coredao/src/transaction.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
+// Source: https://docs.coredao.org/docs/Learn/products/btc-staking/design
+export const CORE_DAO_TESTNET_CHAIN_ID = Buffer.alloc(2, 0x1115);
+export const CORE_DAO_MAINNET_CHAIN_ID = Buffer.alloc(2, 0x1116);
+export const CORE_DAO_SATOSHI_PLUS_IDENTIFIER = Buffer.alloc(4, 0x5341542b);
+
+/** As of v2, this is the construction of the OP_RETURN:
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:6]: move this as a `/* */` comment inside the function body - the `/**` comment should focus on the interface, not the impl

File: modules/utxo-coredao/src/transaction.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ * followed by the length followed by the data (byte[80] -> OP_PUSHDATA + 80 + byte[80])
+ *
+ * Either RedeemScript or Timelock must be available, the purpose is to allow relayer to
+ * obtain the RedeemScript and submit transactions on Core. If a RedeemScript is provided,
+ * relayer will use it directly. Otherwise, relayer will construct the redeem script based
+ * on the timelock and the information in the transaction inputs.
+ *
+ * @returns Buffer OP_RETURN buffer
+ */
+export function createCoreDaoOpReturnBuffer({
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/src/transaction.ts:32]:
```suggestion
export function createCoreDaoOpReturnOutputScript({
```

File: modules/utxo-coredao/test/unit/transaction.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    );
+    assert.throws(() =>
+      createCoreDaoOpReturnBuffer({
+        version: validVersion,
+        chainId: validChainId,
+        delegator: validDelegator,
+        validator: validValidator,
+        fee: validFee,
+      })
+    );
```
Comment (OttoAllmendinger) [inline modules/utxo-coredao/test/unit/transaction.ts:87]: hmm why tho?

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-coredao/test/unit/transaction.ts (hunk 9 above)
Comment (davidkaplanbitgo) [reply modules/utxo-coredao/test/unit/transaction.ts:87]: It needs to have either a redeem script or a timelock

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]
please update codeowners

**Review by therealdwright — APPROVED** [review]

File: modules/utxo-coredao/test/unit/transaction.ts (hunk 9 above)
Comment (OttoAllmendinger) [reply modules/utxo-coredao/test/unit/transaction.ts]: deserves a comment, however let's make it a type error instead

**Review by OttoAllmendinger — APPROVED** [review]

File: modules/utxo-coredao/test/unit/transaction.ts (hunk 9 above)
Comment (OttoAllmendinger) [reply modules/utxo-coredao/test/unit/transaction.ts]: see other comment

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**